Add Helm Chart

desaintmartin opened the pull request:
I am maintaining the kube-dashboard 2.0.0 helm chart (https://github.com/helm/charts/pull/15744, which is pretty much a whole rewrite of the chart with much more sanity), and I would like to move it to the kubernetes/dashboard repository since everything in helm is being decentralized (and github.com/helm/charts.git being slowly deprecated to move to hub.helm.sh) and continue to (co-)maintain it in this repository.

This is a work in progress and a first proposal in order to discuss.

To do after discussion:
- [x] OWNERS for helm chart (separate? same as repository?)
- [x] README
- [x] Others?
- [ ] Create (through CI?) Helm repository to make it available through hub.helm.sh

desaintmartin commented:
/hold

maciaszczykm commented:
`OWNERS` file looks okay, +1 from my side to keep you as one of them.

> Create (through CI?) Helm repository to make it available through hub.helm.sh

What do you need to complete this task?

Would you like to merge the files as it is or to continue and perform changes from 1 PR?

desaintmartin commented:
1/ OWNERS: Great! I am not yet a kubernetes org member, I would probably need sponsorship for this.

2/ Releasing: should we sync dashboard versions and its helm chart versions? If so, we could build and push the helm chart into each github release (can easily be done through GitHub Actions) and generate through CI (Travis? Github Actions?) an index that will be used by the Helm Hub through the gh-pages branch. There's still a question about the testing part to be on par with helm/charts.git, but I guess we can improve over time.
Note that this would probably change the release workflow for the whole project and/or add a new CI through Github Actions.

3/ I think it is better to hold this PR until publishing is ready!

desaintmartin commented:
So I have a working CI that will push to gh-pages. Downside is that you need to create a github token with write permissions and set it in travis configuration. The other way is to commit an encrypted deploy token in the repository... Or yet another way is to do it manually.

floreks commented on .travis.yml:
```diff
@@ -36,22 +36,54 @@ before_script:
         - aio/scripts/install-codegen.sh
       script: npm run check
-    - script: npm run test:coverage
+    - name: "Test coverage"
```
s/Test coverage/Unit test

floreks commented on .travis.yml:
```diff
@@ -36,22 +36,54 @@ before_script:
 jobs:
   include:
     - stage: test
+      name: "Test"
```
I assume this is name for the `npm run check` then
s/Test/Code check & linting

floreks commented on .travis.yml:
```diff
@@ -36,22 +36,54 @@ before_script:
         - npm run docker:push:head
+
     - stage: release
+      name: "Docker push"
```
s/Docker push/Release

floreks commented on .travis.yml:
```diff
@@ -36,22 +36,54 @@ before_script:
+      script: npm run cluster:start && npm run e2e
+
     - stage: deploy
+      name: "docker push head"
```
s/docker push head/Development release

floreks commented:
Mostly it lgtm. The only thing I am worried about is how to properly configure travis to be able to push to the repo and branch `gh-pages`.

desaintmartin commented:
Thanks for the review!

First of all, we DON'T have to automate it now:
- It could be a manual operation done by a maintainer with push access to the repo
- Travis could be configured, but push to ANOTHER repository in order to preserve integrity to kubernetes/dashboard
- The other solution is to maintain a stateful app (namely chartmuseum) that runs somewhere and connects to an object storage (gcs?) to serve Charts. Populated manually or through travis.

That being said, in order to enable the proposed CI, one would have to put into the travis config a GitHub token with push access. This is one huge current limitation of GitHub Pages. This is quite simple to do but have serious implications (Travis is able to push, and whoever has access to Travis parameters could grab the token and... push) so it has to be carefully weighted depending on who has access.

Once again, the PR can be split into two.

floreks commented:
Then I'd move the `gh-pages` logic to a script and make it a manual step that should be executed after creating a new release tag. Keeping token with push access to the repo on travis would be a security risk.

desaintmartin commented:
Here is a first version of the script, taking the tag to add to gh-pages as argument.
Sorry for the force-pushes!

bskim45 commented:
It would be great to add metrics-scraper sidecar as the official manifest suggests. https://github.com/kubernetes/dashboard/tree/master/aio/deploy/recommended

Or should we propose a seperate PR to add metrics-scraper?

desaintmartin commented:
Should we have a Helm Chart in https://github.com/kubernetes-sigs/dashboard-metrics-scraper and make this one depend on it, in a future PR, in order to have clear separation of concern? The only advantage of putting metrics-scraper here would be to have a sidecar in the main Pod.

davidalger commented on aio/deploy/helm-chart/kubernetes-dashboard/values.yaml:
```diff
@@ -0,0 +1,174 @@
+## Serve application over HTTP without TLS
+##
+## Note: If set to true, you may want to add --enable-insecure-login to extraArgs
+protocolHTTP: false
```
Casing of this default value does not match what is currently used throughout the templates.
```suggestion
protocolHttp: false
```

davidalger replied:
Note: I discovered this because I copied it from this file, and it wasn't working as expected until I changed my `--set` to use the properly cased value name.

desaintmartin replied:
Indeed, thanks for noticing, let me fix it and squash it.

desaintmartin commented:
Fixed Helm 3 compatibility.

davidalger commented on aio/deploy/helm-chart/kubernetes-dashboard/README.md:
```diff
@@ -0,0 +1,132 @@
+`podDisruptionBudget.enabled`       | Create a PodDisruptionBudget                                                                                                | `false`
+`podDisruptionBudget.minAvailable`  | Minimum available instances; ignored if there is no PodDisruptionBudget                                                     |
+`podDisruptionBudget.maxUnavailable`| Maximum unavailable instances; ignored if there is no PodDisruptionBudget                                                   |
+`securityContext`                   | PodSecurityContext for pod level securityContext                                                                            | `{allowPrivilegeEscalation:false, readOnlyRootFilesystem: true, runAsUser: 1001, runAsGroup: 2001}`
```
Might want to update this as it no longer reflects the contents of `values.yaml`

desaintmartin replied:
done! Thanks.

davidalger commented on aio/deploy/helm-chart/kubernetes-dashboard/values.yaml:
```diff
@@ -0,0 +1,168 @@
+## SecurityContext for the kubernetes dashboard container
+dashboardContainerSecurityContext:
+  allowPrivilegeEscalation: false
+  readOnlyRootFilesystem: true
```
This should be updated to reflect moving the entire contents of `securityContext` to `dashboardContainerSecurityContext`:

```yaml
## PodSecurityContext for pod level securityContext
securityContext: {}

## SecurityContext for the kubernetes dashboard container
dashboardContainerSecurityContext:
  allowPrivilegeEscalation: false
  readOnlyRootFilesystem: true
  runAsUser: 1001
  runAsGroup: 2001
```

reference: https://github.com/kubernetes/dashboard/blob/master/aio/deploy/recommended/06_dashboard-deployment.yaml#L62-L66

desaintmartin replied:
Solved! Thanks.

davidalger replied:
ty! :)

davidalger commented:
@bskim45 @desaintmartin Opened a PR to the branch this PR is based on which adds support for the `dashboard-metrics-scraper` and would appreciate any feedback you might have. I've pretty much based the effort on the recommended deployment found here: https://github.com/kubernetes/dashboard/tree/master/aio/deploy/recommended

davidalger commented on aio/deploy/helm-chart/kubernetes-dashboard/templates/clusterrole-metrics.yaml:
```diff
@@ -0,0 +1,13 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: "{{ template "kubernetes-dashboard.fullname" . }}"
```
@desaintmartin I'm trying to wrap my head around why an additional clusterrole and clusterrolebinding were added with their own settings. This conflicts (it's identical to) what's in clusterrole.yaml and attempts to deploy the latest here now fail because with rbac.create=true and rbac.clusterRoleMetrics=true the chart is attempting to create duplicate clusterrole and clusterrolebinding resources of the same name.

```
clusterroles.rbac.authorization.k8s.io "kubernetes-dashboard" already exists
```

Were these changes related to the PR I opened with support for the dashboard-metrics-scraper or something else?

desaintmartin replied:
I've just renamed clusterrole.yaml to clusterrole-metrics.yaml and clusterrolebinding.yaml to clusterrolebinding-metrics.yaml, and be more modular about which one(s) you want.

Previously, you had to install all 4 at once (all or nothing).

So this must be related to to this change. Does this resource still exist if you uninstall the release? I am not able to reproduce on my cluster (which does not mean it is not my fault!).

davidalger replied:
Ok, so this was totally my fault. Had used `cp -R` to copy in changes from the updated PR here, and ended up with both the old files _and_ the new ones in my local copy of the chart.

Sorry for the false alarm! 🙈

desaintmartin commented:
@davidalger: I propose we include the metrics once we have a first stable version of the chart.

We have to decide whether kubernetes/dashboard.git includes things about metrics-server, or if metrics-server.git should have its own helm chart and kubernetes-dashboard (optionally) depends on it. I tend to prefer the second solution in order to reduce coupling.

What do you think?

(BTW: updated to 2.0.0-rc1)

davidalger commented:
@desaintmartin Leaving a crucial component of the dashboard out of the first release of the chart IMO cripples it and puts it at odds with the recommended deployment of `kubernetes-dashboard` which includes the `dashboard-metrics-scraper` (an important part of having a fully functional dashboard)

You are correct, the README is a little bit behind and still references the deprecated use of heapster (which is still supported by v2, as it'll use it if present to display metrics as well as the bundled `dashboard-metrics-scraper` but no longer preferred and is unlikely to be used in new deployments). This wouldn't stop me from including the scraper in the initial release of the chart however, and cloud providers such as AWS are already building their docs to have two steps: 1) deploy metrics-server; 2) deploy kubernetes-dashboard; to setup the Kubernetes UI: https://docs.aws.amazon.com/eks/latest/userguide/dashboard-tutorial.html The docs here in this repository need a few updates it seems to fully cover the changes in dashboard v2, both the README and the integrations wiki page: https://github.com/kubernetes/dashboard/wiki/Integrations as they will be out of date once v2 hits GA.

> or if metrics-server.git should have its own helm chart and kubernetes-dashboard (optionally) depends on it. I tend to prefer the second solution in order to reduce coupling.

Metrics server would not belong bundled into this chart. In some cases it will already be deployed by default (per [Kubernetes docs](https://kubernetes.io/docs/tasks/debug-application-cluster/resource-metrics-pipeline/): It is deployed by default in clusters created by `kube-up.sh` script as a Deployment object). When it's not there, it's as simple as either deploying from the manifests in metrics-server repository (which is what AWS EKS docs is suggesting users do), or leveraging the currently available chart for it (as I am in the cluster I'm currently building):

```
helm install  metrics-server --namespace kube-system stable/metrics-server
```

Is the lack of updated documentation on the use of metrics-server vs heapster the only blocker from your perspective? FWIW, the dashboard will function without metrics-scraper installed, even when dashboard-metrics-scraper is included, it just won't show the metrics data until the user installs the metrics-server.

davidalger commented on aio/deploy/helm-chart/kubernetes-dashboard/values.yaml:
```diff
@@ -0,0 +1,176 @@
+    cpu: 100m
+    memory: 100Mi
+  limits:
+    cpu: 10
```
Should this really be set to `10` cpu core limit by default? Or is this perhaps a typo? Noticed my 2-core t3.medium was way committed, and it boiled down to this pod having a limit of 10. Seems terribly high.

desaintmartin replied:
This is a typo, thank you very much.
I would suggest a default of 2 cpus (more than 1, in order not to be throttled by kernel to 1.00 and sometimes a bit less due to approximations of CFS.

Note that Kubernetes good practices tend to go to NO limits.cpu. I tend to be on the middle : set the limit high enough so that it is not a problem, but still save the VM if a pod goes rogue.

davidalger replied:
2 sounds reasonable to me, and I agree NO limits.cpu is not likely to be the best course of action for most Pods including this one.

desaintmartin commented:
@davidalger could you give me the URL of you PR?

davidalger commented:
@desaintmartin PR can be found here: https://github.com/wiremind/dashboard/pull/1

I made that PR against the branch this PR was created from, so if merged this PR would update and include the additional templates.

davidalger commented on aio/deploy/helm-chart/kubernetes-dashboard/templates/networkpolicy.yaml:
```diff
@@ -0,0 +1,38 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+{{- if .Values.networkPolicy -}}
```
```suggestion
{{- if .Values.networkPolicy }}
```
The chomp on this will cause a rendering error when networkPolicy is enabled.

davidalger commented on aio/deploy/helm-chart/kubernetes-dashboard/templates/pdb.yaml:
```diff
@@ -0,0 +1,32 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+{{- if .Values.podDisruptionBudget.enabled -}}
```
```suggestion
{{- if .Values.podDisruptionBudget.enabled }}
```
Chomp on this will cause an error when enabled.

davidalger commented on aio/deploy/helm-chart/kubernetes-dashboard/templates/serviceaccount.yaml:
```diff
@@ -0,0 +1,23 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+{{- if .Values.serviceAccount.create -}}
```
```suggestion
{{- if .Values.serviceAccount.create }}
```
Causes following error due to chomp basically moving `apiVersion` into the comment text:
```
Error: unable to build kubernetes objects from release manifest: error validating "": error validating data: apiVersion not set
```
```

maelvls replied:
Here is what the result looks like:

```yaml
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    app.kubernetes.io/name: kubernetes-dashboard
```

maelvls replied:
I think the actual problem is the `{{-` at the beginning. Since the top of the file has comments, we should remove all the whitespaces before `{{-`. What would work is:

```suggestion
{{ if .Values.serviceAccount.create -}}
```

desaintmartin replied:
Seems to work with the second `-` removed. Thank you very much!

maelvls commented on aio/deploy/helm-chart/kubernetes-dashboard/README.md:
````diff
@@ -0,0 +1,133 @@
+## TL;DR
+
+```console
+helm repository add kubernetes-dashboard https://https://kubernetes.github.io/dashboard/
````
```suggestion
helm repository add kubernetes-dashboard https://kubernetes.github.io/dashboard/
```

desaintmartin replied:
Thanks!

sciyoshi commented on aio/deploy/helm-chart/kubernetes-dashboard/templates/ingress.yaml:
```diff
@@ -0,0 +1,67 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+{{- if .Values.ingress.enabled -}}
```
```suggestion
{{ if .Values.ingress.enabled -}}
```

Without this, the rendered template looks like this:

```
# See the License for the specific language governing permissions and
# limitations under the License.apiVersion: extensions/v1beta1
kind: Ingress
metadata:
```

and helm complains about a missing `apiVersion`.

desaintmartin replied:
Indeed. Adding the licence broke all those files. I'll work to have better testing.

sciyoshi commented on aio/deploy/helm-chart/kubernetes-dashboard/templates/ingress.yaml:
```diff
@@ -12,7 +12,7 @@
 # limitations under the License.
 
-{{- if .Values.ingress.enabled -}}
+{{- if .Values.ingress.enabled }}
```
This fix doesn't work, because the whitespace is still removed by the next tag. The correct fix is:
```suggestion
{{ if .Values.ingress.enabled -}}
```

desaintmartin replied:
I'm going to add several tests in CI later today. It should solve once and for all those issues.

desaintmartin replied:
Here we are, along with the correct fixes: https://travis-ci.org/wiremind/dashboard/jobs/645872825

atoato88 commented on aio/deploy/helm-chart/kubernetes-dashboard/templates/clusterrolebinding-metrics.yaml:
```diff
@@ -0,0 +1,30 @@
+  name: {{ template "kubernetes-dashboard.fullname" . }}
+subjects:
+  - kind: ServiceAccount
+    name: kubernetes-dashboard
```
In other places used `ServiceAccount`(e.g. [here1](https://github.com/kubernetes/dashboard/pull/4502/files#diff-ed120a3d919cabbd8ade3ae964f3a424R51), [here2](https://github.com/kubernetes/dashboard/pull/4502/files#diff-457ad80f5b951b1fd5851a488d2b7e90R28), and so on), `name` is set to `{{ template "kubernetes-dashboard.serviceAccountName" . }}`. Isn't it needed here also?

vutny replied:
```suggestion
    name: {{ template "kubernetes-dashboard.serviceAccountName" . }}
```

I can confirm, it is definitely should be like this, otherwise metrics scrapping would not work with following sympthoms:
```
$ kubectl -n monitoring logs dashboard-metrics-scraper-5c7db799bf-2hcrf
192.168.99.248 - - [20/Mar/2020:11:12:12 +0000] "GET / HTTP/1.1" 200 6 "" "kube-probe/1.14+"
{"level":"error","msg":"Error scraping node metrics: nodes.metrics.k8s.io is forbidden: User \"system:serviceaccount:monitoring:dashboard-kubernetes-dashboard\" cannot list resource \"nodes\" in API group \"metrics.k8s.io\" at the cluster scope","time":"2020-03-20T11:12:21Z"}
```

and 

```
$ kubectl -n monitoring logs dashboard-kubernetes-dashboard-649dffc9dc-6vgxh
2020/03/20 10:53:39 Skipping metric because of error: Metric label not set.
```

desaintmartin replied:
I've re-worked how metrics-scraper behaves and fixed it. Thanks!

atoato88 commented on aio/deploy/helm-chart/kubernetes-dashboard/README.md:
```diff
@@ -0,0 +1,133 @@
+`podDisruptionBudget.maxUnavailable`| Maximum unavailable instances; ignored if there is no PodDisruptionBudget                                                   |
+`securityContext`                   | PodSecurityContext for pod level securityContext                                                                            | `{}`
+`dashboardContainerSecurityContext` | SecurityContext for the kubernetes dashboard container                                                                      | `{allowPrivilegeEscalation:false, readOnlyRootFilesystem: true, runAsUser: 1001, runAsGroup: 2001}`
+`networkPolicy`                     | Whether to create a network policy that allows access to the service                                                        | `false`
```
Use same format along with other parameters.

```suggestion
`networkPolicy.enabled`                     | Whether to create a network policy that allows access to the service                                                        | `false`
```

desaintmartin replied:
Thanks!

atoato88 commented on aio/deploy/helm-chart/kubernetes-dashboard/README.md:
```diff
@@ -0,0 +1,133 @@
+`image.tag`                         | Image tag                                                                                                                   | `v2.0.0`
+`image.pullPolicy`                  | Image pull policy                                                                                                           | `IfNotPresent`
+`image.pullSecrets`                 | Image pull secrets                                                                                                          | `[]`
+`annotations`                       | Annotations for deployment                                                                                                  | `{}`
```
There is not description about `labels` here.
It maybe following?

```suggestion
`annotations`                       | Annotations for deployment                                                                                                  | `{}`
`labels`                       | Labels for deployment                                                                                                  | `{}`
```

desaintmartin replied:
Thanks! I've added a few missing values in README, including this one.

rifelpet commented on aio/deploy/helm-chart/kubernetes-dashboard/templates/ingress.yaml:
```diff
@@ -0,0 +1,67 @@
+{{- $serviceName := include "kubernetes-dashboard.fullname" . -}}
+{{- $servicePort := .Values.service.externalPort -}}
+{{- $paths := .Values.ingress.paths -}}
+apiVersion: extensions/v1beta1
```
```suggestion
apiVersion: networking.k8s.io/v1beta1
```

This affects the [1.X chart](https://github.com/helm/charts/blob/master/stable/kubernetes-dashboard/templates/ingress.yaml#L5) as well. Would a PR to the old chart be accepted?

desaintmartin replied:
Since this API has been introduced in kube 1.14, should we support older kube versions? If not, let's do this, if so, we can add a condition to use either the old one or the new one? What do you think?

rifelpet replied:
You could support both with a conditional like this: https://github.com/helm/charts/blob/2897f04b79ee03487d9a995ea9b8f0618953cbca/stable/prometheus/templates/server-ingress.yaml#L7-L11

I think that would be sufficient. And dont worry about my question regarding the old chart, we ended up upgrading to this one :)

rifelpet commented on aio/deploy/helm-chart/kubernetes-dashboard/README.md:
```diff
@@ -0,0 +1,133 @@
+- It enables by default values for `podAnnotations` and `securityContext`, please disable them if you don't supoprt them
+- It removes `enableSkipLogin` and `enableInsecureLogin` parameters. Please use `extraEnv` instead.
+- It adds a `ProtocolHttp` parameter, allowing you to switch the backend to plain HTTP and replaces the old `enableSkipLogin` for the network part.
+- If `ProtocolHttp` is not set, it will automatically add to the `Ingress`, if enabled, annotations to support HTTPS backends for nginx-ingress and GKE Ingresses.
```
```suggestion
- If `protocolHttp` is not set, it will automatically add to the `Ingress`, if enabled, annotations to support HTTPS backends for nginx-ingress and GKE Ingresses.
```

desaintmartin replied:
Thanks!

desaintmartin commented:
Hello @davidalger, after our discussion, I've added your commits regarding metrics-scraper into the PR. I wonder if we should add metrics-server as an (optional) dependency as well. What do you think?
Other people, for advice: metrics-scraper deployment is now enabled by default, is that an expected default value to get started ?

davidalger commented:
I signed it

davidalger commented:
> I wonder if we should add metrics-server as an (optional) dependency as well. What do you think?

I think that would be a good idea perhaps. I've not personally used optional dependencies, but if I can find the time to, I'll take a look at what it might look like.

Since the docs state a dependency on either heapster or metrics-server, it would definitely need to be an optional one to avoid breaking the ability for users to deploy the new dashboard into older clusters still running heapster.

vutny commented:
@desaintmartin I'm leaving a link to review comment here since it might get lost in the long thread:
https://github.com/kubernetes/dashboard/pull/4502#discussion_r395591005

For anybody who's testing this, take a look on above suggestion if you see no CPU/memory metrics and those messages in the logs:
```
Skipping metric because of error: Metric label not set.
```
```json
{"level":"error","msg":"Error scraping node metrics: nodes.metrics.k8s.io is forbidden: User \"system:serviceaccount:monitoring:dashboard-kubernetes-dashboard\" cannot list resource \"nodes\" in API group \"metrics.k8s.io\" at the cluster scope","time":"2020-03-20T11:47:21Z"}
```

desaintmartin commented:
So, I've adapted the dashboard-metrics-scraper to make it a sidecar container, added an optional metric-server sub-chart and cleaned up a bit README. I've squashed the older commits for sanity.

It can be still complex to have a working metrics-server + dashboard-metrics-scraper, here is my configuration:
```yaml
rbac:
  create: true
  clusterRoleMetrics: true

metricsScraper:
  enabled: true

metrics-server:
  enabled: true
  args:
    - --kubelet-preferred-address-types=InternalIP
    - --kubelet-insecure-tls
```

But I guess this can be improved in the future. What do you think about those improvements?

plnordquist commented on aio/deploy/helm-chart/kubernetes-dashboard/requirements.yaml:
```diff
@@ -0,0 +1,5 @@
+  - name: metrics-server
+    version: 2.x.x
+    repository: https://kubernetes-charts.storage.googleapis.com/
+    condition: metricsServer.enabled
```
This condition doesn't match the values file or the prefix for other configuration options.  Seems like either the dependency should have an alias or the condition should be changed to `metrics-server.enabled`.

desaintmartin replied:
Indeed, thanks, fixed!

davidalger commented:
@desaintmartin Don't know that I have a strong opinion either way on making the metrics-scraper run as a sidecar container rather than an independent/sidecar pod. It works either way, but is perhaps more flexible running it out as an additional deployment. This method keeps it running as a single deployment unit so scaling up one requires scaling up the other. Doesn't look like there was any discussion about this before it was added to the manifests for example deployment, it was just merged as a sidecar pod rather than sidecar container: https://github.com/kubernetes/dashboard/pull/3504

There is an issue with these changes and I'm honestly not sure what is causing it, but it's related to specifying the dependency on metrics-server in the requirements.yaml file.

I uninstalled the previous version of this chart from my cluster and when I attempted to install, I got the following error:

```
$ helm -n kube-system install kubernetes-dashboard ./charts/kubernetes-dashboard -f /tmp/values.yml
Error: found in Chart.yaml, but missing in charts/ directory: metrics-server
```

If I comment out the contents of `requirements.yaml` the chart deploys successfully and the metrics-scraper works in tandem with the metrics-server (which is deployed separately in my case). This is what my values file looks like:

```
extraArgs:
- --enable-insecure-login
protocolHttp: true
rbac:
  clusterReadOnlyRole: true
service:
  externalPort: 8080
metricsScraper:
  enabled: true
```

atoato88 commented:
@desaintmartin 
There are 2 items to action.

1. You need to amend commit message to take into account message rule, that's why CI error occurred.
Please see `k8s-ci-robot` [comment](https://github.com/kubernetes/dashboard/pull/4502#issuecomment-607658226) for detail.

2. And, you need to be member of the kubernetes org.
Plsease see [comment here](https://github.com/kubernetes/dashboard/pull/4502#issuecomment-607658457) and [liked page](https://github.com/kubernetes/community/blob/master/community-membership.md#member).


Please follow up this by other members !! :smile:

desaintmartin commented:
thanks @atoato88, indeed I haven't progressed on the "kube org member" part yet. Actually I would need another sponsor but I haven't really looked at it yet. If anybody can sponsor me... 
Regarding the signoff commit error, since this is on /hold, I've kept temporary commits, so that people can see the changes, that are NOT meant to be merged but to be squashed before merge (unless the merge system automatically squashes...).


@davidalger oh, actually you need to do `helm dependency build` in order to download the metrics-server as a tgz inside of `./charts` directory. It is not a good practice to commit it so it is actually part of the "release" script. Once it will be uploaded and served, of course a simple `helm install my-dashboard kubernetes-dashboard/kubernetes-dashboard` will work out of the box.

Regarding the pod vs sidecar container, I've observed that in a 2500-pods cluster:
- dashboard-metrics-scraper with default configuration uses ~150MB of memory with negligible cpu
- metrics-server use, for each replica of kube-dashboard/dashboard-metrics-scraper, an additional 50Kb/s of network and 0.01 cpu (and negligible additional memory).

Extrapolating for those figures, a cluster running 10 replicas of this would not put metrics-server in danger and would use approximately 1.5GB more memory than the autonomous deployment/service approach (if it increases linearly...). On the other hand, it eases maintenance and simplify the logic (one less deployment, one less service to be exposed).

Even if I tend to prefer the sidecar approach, it is really easy to rollback to use the standalone approach! @jeefy, what is your opinion on this? Should we assume a very high number of replicas and isolate scraper from dashboard?

davidalger commented:
@desaintmartin Aha, that is what I was missing. Thanks! Works perfectly (even with metrics-server installed separately) once I did that. Appreciate it.

Those numbers and logic behind the pattern change feel solid to me as well. If one is running 10 replicas of this to distribute load on the dashboard, probably want to distribute the load on the dashboard scraper in a similar fashion. I do like that it has fewer moving pieces this way. I'm newer in the k8s ecosystem though, so take what I say there with a grain of salt. My exp is limited to much smaller clusters thus far. :)

pierluigilenoci commented:
What about this? https://github.com/funkypenguin/helm-kubernetes-dashboard

funkypenguin commented:
@pierluigilenoci the above is my fork of https://github.com/wiremind/dashboard/tree/helm-chart, it's 100% aligned with this one. I published it on the basis that I want to use the chart in my CI (which requires a published repo) which is most likely to end up "official". I've PR'd some of my fixes back to https://github.com/wiremind/dashboard/tree/helm-chart. You could use it in the interim until this is merged, but once this is merged I'll be deprecating my chart in favor of this one.

desaintmartin commented:
/hold cancel

Eddman commented:
I've tested that chart and found out that as @funkypenguin wrote there are wrongly placed security contexts and resources. Appart of that there is:
```
metricsScraper:
  revisionHistoryLimit: 10
```
That one is never used.

Also the metrics scraper sidecar has wrong version `1.0.3`. Current is `1.0.4`.

Can we get this fixed and release the chart asap? Would be nice to have official chart that we can download and use. Right now I have my own local copy...

floreks commented:
@Eddman thanks for testing. If it will be ready I'll also give it a final check and we can merge.

desaintmartin commented:
Thank you both @Eddman and @funkypenguin, I've updated the PR to reflect this.

Eddman commented:
@desaintmartin Checked your changes. 

One thing I don't like is the `ContainerSecurityContext` in `metricsScraper`. [Helm best practice docs](https://helm.sh/docs/chart_best_practices/values/) the naming conventions are saying: always camelCase, no initial caps.

Also I'd suggest to rename both `dashboardContainerSecurityContext` and `metricsScraper.ContainerSecurityContext` to `securityContext`/`metricsScraper.securityContext`. This is convention used in almost all other charts.

desaintmartin commented:
Thanks for the check @Eddman, the first one was a typo, and I hesitated for the second one so that it is explicit but sure, let me change that.

Eddman commented:
@desaintmartin Thanks for quick response! 

`containerSecurityContext` vs. `securityContext` - I mean both are valid. I'd only say that it become somehow community standard to use the shorter `securityContext`, but it is nothing I want to fight over.

One small think: in README.md we still have ref to `dashboardContainerSecurityContext`.

davidalger commented:
@Eddman The two values here `containerSecurityContext` and `securityContext` are for two separate `securityContext` values. The `securityContext` value applies to the Pod and `containerSecurityContext` applies to the container within the Pod.

I've seen some charts use `podSecurityContext` and `securityContext` (for the container) and others use `podSecurityContext` and `containerSecurityContext` to avoid the possible confusion. Perhaps it would be good for charts to move towards the latter convention to help clear things up.

Eddman commented:
Ok, it makes sense, thanks.

desaintmartin commented:
Fixed README.

Eddman commented:
@desaintmartin Super stupid thing: I've noticed that all files have LICENSE: `# Copyright 2017` Maybe worth changing to 2020 :man_shrugging: ?

floreks commented:
No need to update the copyright date. It should reflect the original date of creation, not get updated every year.

Eddman commented:
Agree, but the files aren’t even created, yet. These are/will be copletely new in this repo.

floreks commented:
Those nuances are always tricky and I am no lawyer :slightly_smiling_face:. It could be explained in many different ways IMHO. Files definitely are already created somewhere, even if they are just in a PR from your work, not "directly" on this repo. They have been created some time ago and possibly we should reflect the original creation date, not the date of "copying" them over here.

Eddman commented:
Did some check and compared with the original [helm/charts/stable/kubernetes-dashboard](https://github.com/helm/charts/tree/master/stable/kubernetes-dashboard) and non-of these files had any copyright before and some of the files with year 2017 are new in this chart :man_shrugging: :smile:

desaintmartin commented:
Anyway, CI will check that we have the "2017" copyright: https://github.com/kubernetes/dashboard/blob/master/aio/scripts/license-header.txt

If we have to change the year, we need to change the CI script as well!

Eddman commented:
Further more the chart was funded in 2017, however some of the files were created in 2018 (by me for instance, helm/charts/pull/3935) and some in 2019. :wink:

Eddman commented:
@desaintmartin So even if you create a file now, you have to add `Copyright 2017`? :thinking: 

But I don't fight here, it's just something that I've spotted there and is not important at all. The chart itself looks good to me, I haven't found any other problems/issues. We have it installed in our test cluster for week now, including the new changes.

floreks commented:
@Eddman good point. We should not check for the year in the license header or extend the check to only check if the year is between 2017 and the current year.

desaintmartin commented:
As @funkypenguin points out in another PR, we don't have a readiness probe (because from dashboard point of view, liveness equals readiness). Should we add a readiness probe duplicating the liveness probe in order to please some validators? Should we keep it like this (and in this case, kubernetes considers readiness from the liveness probe)?

(Note that this can be postponed to a future and separate PR since the helm chart follows the official deployment files regarding this topic.)

iAnomaly commented:
@desaintmartin Can you help me understand why you removed the namespace declaration in the secret  and service account with 09405553ad94b880cac7f1f56c814e24970ac5b7?

I'm attempting to use your chart directly until it is merged and this change breaks out of the target namespace encapsulation boundary by attempting to create the secret and service account in the default namespace.

desaintmartin commented:
> @desaintmartin Can you help me understand why you removed the namespace declaration in the secret and service account with [0940555](https://github.com/kubernetes/dashboard/commit/09405553ad94b880cac7f1f56c814e24970ac5b7)?
> 
> I'm attempting to use your chart directly until it is merged and this change breaks out of the target namespace encapsulation boundary by attempting to create the secret and service account in the default namespace.

Sure, this was a legacy from old days, and helm does NOT need namespace to be specified. Could you specify your helm version and the command used to install, so that I try to reproduce? Is that an upgrade or an install? I have running Releases of this PR in clusters, and there is no trace of Secret or ServiceAccount outside of its namespace.

iAnomaly commented:
> > @desaintmartin Can you help me understand why you removed the namespace declaration in the secret and service account with [0940555](https://github.com/kubernetes/dashboard/commit/09405553ad94b880cac7f1f56c814e24970ac5b7)?
> > I'm attempting to use your chart directly until it is merged and this change breaks out of the target namespace encapsulation boundary by attempting to create the secret and service account in the default namespace.
> 
> Sure, this was a legacy from old days, and helm does NOT need namespace to be specified. Could you specify your helm version and the command used to install, so that I try to reproduce? Is that an upgrade or an install? I have running Releases of this PR in clusters, and there is no trace of Secret or ServiceAccount outside of its namespace.

Got it. Thank you. I didn't know Helm did this. I was confused because I saw the namespace line disappear from those templates during a Helm diff during a chart upgrade to your version, however after inspecting the ServiceAccount I do see it is still located in the expected namespace and contains the namespace declaration in its internal manifest. Helm must even do this after its own plugin pipeline.

Thanks for putting together this chart and submitting the PR. Sure hope it can be merged soon, but until then I'll use your fork. Cheers!

desaintmartin commented:
@floreks so, should we change the licence check script to allow any year, and change them accordingly? Or should we keep it as it is now?
Once this is settled, I think we can merge (and of course do other non-breaking PRs thereafter) now that I can be OWNER of the chart.

floreks commented:
I guess we can update license check later.

Eddman commented:
Are we waiting here for something? Or can we already merge it and release the chart?

rifelpet commented:
Has thought been given to how the chart will be hosted? looking at the [helm hub repos](https://github.com/helm/hub/blob/master/config/repo-values.yaml) it seems that most host it on github.io using [chart-releaser](https://github.com/helm/chart-releaser). Will that process be automated or manually run by a maintainer?

desaintmartin commented:
@rifelpet
chart-releaser works (or, at least, worked, last time i've checked) best with several charts. Since we have only one, a maintainer should just run the `aio/scripts/helm-release-chart.sh` (along with its friend `aio/scripts/helm-lint.sh` run by ci) to generate a new chart version, then push to the gh-pages branch of this repository, which would make the helm index accessible through `https://kubernetes.github.io/dashboard` which will be added to the Helm Hub (https://github.com/helm/hub/blob/master/Repositories.md).

Automating this would require a r/w token to be created and given to a CI/CD system (being travis, github actions, prow or any other one), and it has been decided it is not wise exposing this repository this way.

Alternatively, we could create a dedicated github repository where only the helm charts are pushed. Last time I discussed, it seemed better to keep k/dashboard and have manual pushes.

llamahunter commented:
what's the status of this?  Any chance this chart will be officially released soon?

ReSearchITEng commented:
> what's the status of this? Any chance this chart will be officially released soon?
In meanwhile you may use it from this repo:
helm repo add funkypenguin-kubernetes-dashboard https://funkypenguin.github.io/helm-kubernetes-dashboard/
(from https://github.com/funkypenguin/helm-kubernetes-dashboard )
It's the same chart as the one expected to be merged here).

aitorhh commented:
The helm chart provided does not include all the [deploy/recommented](https://github.com/wiremind/dashboard/tree/helm-chart/aio/deploy/recommended) resources. It is missing the [scrapper](https://github.com/wiremind/dashboard/blob/helm-chart/aio/deploy/recommended/08_scraper-deployment.yaml). 

Should that be part of the helm chart as well? Since `metrics-servers` is a dependency, I believe it makes sense to add it, at least as configurable option.

pierluigilenoci commented:
@aitorhh I disagree. The chart should contain the bare minimum. [1] If you want to install metrics-server use the metrics-server chart [2]. It hasn't been migrated away from stable yet but it doesn't matter, it's the idea what really matters. If you want you can create an umbrella chart with all you need [1].

[1] https://helm.sh/docs/howto/charts_tips_and_tricks/#complex-charts-with-many-dependencies
[2] https://github.com/helm/charts/tree/master/stable/metrics-server

ReSearchITEng commented:
@aitorhh Isn't this what you are looking for?
https://github.com/wiremind/dashboard/blob/helm-chart/aio/deploy/helm-chart/kubernetes-dashboard/values.yaml#L139

aitorhh commented:
@ReSearchITEng , I was wondering if the metric-scraper should be part of the dashboard helm chart or a dependency. But as @pierluigilenoci pointed out, it might be better to have it as dependency. However at the moment there is not offical chart for that.

desaintmartin commented:
Please see the values.yaml of this PR for informations about how to enable dashboard-metrics-scraper and metrics-server.
I try to maintain https://wiremind.github.io/dashboard/ as a repository mirroring this PR.

pierluigilenoci commented:
@aitorhh sorry but I read wrong. IMHO the sidecar of the scraper makes sense that it is on the chart in some way, the metrics-server not. It would be like adding in a chart of software that exposes an endpoint for Prometheus the installation of Prometheus itself. Or vice versa, add the installation of Prometheus to the Grafana chart or any other software that consumes the data.

desaintmartin commented:
metrics-server is an optional [subchart](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/), see requirements.yaml, so it respects the separation of concern and all possible use cases. ;)

maciaszczykm commented:
@desaintmartin hey, could you share the status here? Is it far away from getting merged?

Eddman commented:
@desaintmartin Hi, one more thing that can be done: Please change `Ingress` API version to:
```
{{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1beta1" }}
apiVersion: networking.k8s.io/v1beta1
{{ else }}
apiVersion: extensions/v1beta1
```
to support new `networking.k8s.io/v1beta1` API version (since 1.14). `extensions/v1beta1` is deprecated and will be removed.

desaintmartin commented:
`extensions/v1beta1` ingresses will be removed in 1.22, so I think we still have time to switch in a PR done later. ;)

desaintmartin commented:
@maciaszczykm we can merge when a core maintainer is ready to release (i.e run the helm-release script and push to gh-pages after merge).

maciaszczykm commented:
Sounds good. Remind us if we will not do it this week. Thanks again for all the work done here :)

Eddman commented:
@desaintmartin More I'm thinking about it I'm more convinced that we should just go and use `networking.k8s.io/v1beta1` only. It is supported since `1.14` and Dashboard 2.0+ is meant for newer K8s releases only 🤷‍♂️

Eddman commented:
And, yes I agree that it can be done later, but this chart is the last one in our cluster that raised a flag that it uses a deprecated API version... 😄

desaintmartin commented:
I would rather freeze this PR so that it gets merged, unless newly discovered bug, and then new PRs can be done to change the chart.

Moreover, regarding this topic, although it looks interesting to use newer APIs, it would not improve the PR (no new feature, no bug fix) but would arbitrarily break older versions. I propose we change to the new API in a few months only (like when 1.20/1.21 is out and before 1.22 alpha) unless specific need.

maciaszczykm commented:
/verify-owners

pierluigilenoci commented:
At this point does it make sense to update the chart with the new release?
https://github.com/kubernetes/dashboard/releases/tag/v2.0.1

Jasstkn commented:
Hi, everyone. Is there any ETA for this PR? I can't figure out how many questions are waiting for resolution...

maciaszczykm commented:
I'll try to review this Thursday, totally forgot about it last week.

maciaszczykm commented:
It would be good to update it to v2.0.1.

desaintmartin commented:
I've upgraded to v2.0.1 and bumped chart version to v2.0.1. I propose we try to keep synchronized the versions for now, but I doubt we can do it in the long run.

maciaszczykm commented:
@desaintmartin In our docs directory we have a short manual with the release process, you could add the update of these charts there.

desaintmartin commented:
Done, and updated to 2.0.1.

maciaszczykm commented on .travis.yml:
```diff
@@ -52,22 +52,51 @@ before_script:
 jobs:
   include:
     - stage: test
+      name: "Code check & linting"
```
👍

maciaszczykm commented on .travis.yml:
```diff
@@ -52,22 +52,51 @@ before_script:
       script:
         - docker login -u $DOCKER_RELEASE_USER -p $DOCKER_RELEASE_PASS
         - npm run docker:push
+    # Manual step for now since it requires to add a GitHub Token to Travis which would be a security issue
```
We have already tokens for Docker, Sauce etc. There is even already `GH_TOKEN` you can try if this is valid token already. If not we can add something that will enable this step :)

desaintmartin replied:
Nice!
However I propose we merge this without change, manually release using the current procedure, then automate in a follow-up PR, what do you think?

maciaszczykm replied:
Ok, sure.

maciaszczykm commented on aio/deploy/helm-chart/kubernetes-dashboard/.helmignore:
```diff
@@ -0,0 +1,38 @@
+# This supports shell glob matching, relative path matching, and
+# negation (prefixed with !). Only one pattern per line.
+.DS_Store
+# Common VCS dirs
```
Put an empty line before comments section.

maciaszczykm commented on aio/deploy/helm-chart/kubernetes-dashboard/.helmignore:
```diff
@@ -0,0 +1,38 @@
+*.bak
+*.tmp
+*~
+# Various IDEs
```
Put an empty line before comments section.

maciaszczykm commented on aio/deploy/helm-chart/kubernetes-dashboard/Chart.yaml:
```diff
@@ -0,0 +1,30 @@
+
+apiVersion: v1
+name: kubernetes-dashboard
+version: 2.0.1
```
Thanks for updating to the new version.

maciaszczykm commented on aio/deploy/helm-chart/kubernetes-dashboard/README.md:
````diff
@@ -0,0 +1,143 @@
+
+```console
+helm repo add kubernetes-dashboard https://kubernetes.github.io/dashboard/
+helm install kubernetes-dashboard/kubernetes-dashboard --name my-release
````
Can you comment what `my-release` means? Just a few words below the commands.

maciaszczykm commented:
Overall looks good. I've added a few comments.

sc250024 commented:
When will this get merged? Would be lovely to have!

desaintmartin commented:
This is my fault... I actually forgot to push to the good remote repo.

desaintmartin commented:
The chart is now available from https://kubernetes.github.io/dashboard/, as described in the chart readme. 

I'm going to get this added to the helm hub and deprecate the dashboard chart from the "stable" helm chart repo. 
This might be delayed by a few hours / days because of my newborn baby which is two days old. ;)

desaintmartin commented:
Note that I encourage everybody who had suggestions that could not be included into this PR to open issues / PRs!

pierluigilenoci commented:
@desaintmartin @maciaszczykm you make my day with this PR merged! Thank you!
Me now: https://media.giphy.com/media/mbTaFQPK6iqEo/giphy.gif

HA07Z7XE commented on aio/deploy/helm-chart/kubernetes-dashboard/README.md:
```diff
@@ -0,0 +1,145 @@
+
+- This version upgrades to kubernetes-dashboard v2.0.0 along with changes in RBAC management: all secrets are explicitely created and ServiceAccount do not have permission to create any secret. On top of that, it completely removes the `clusterAdminRole` parameter, being too dangerous. In order to upgrade, please update your configuration to remove `clusterAdminRole` parameter and uninstall/reins ...
+- It enables by default values for `podAnnotations` and `securityContext`, please disable them if you don't supoprt them
+- It removes `enableSkipLogin` and `enableInsecureLogin` parameters. Please use `extraEnv` instead.
```
@desaintmartin do you have an example of how to set these with `extraEnv` ?

desaintmartin replied:
This is a mistake, please use `extraArgs`!
